woocommerce/woocommerce-ios
# [Mobile Payments] Reset Stripe Terminal SDK when logging out and/or switching stores

Closes #4111 

⚠️ This PR is against the feature branch implementing support for Card Present Payments, not against develop ⚠️
⚠️ Testing this PR requires a hardware reader and uploading a pre-release version of WCPay to a test site⚠️

This PR starts where #4140 ends. As pointed out [in this comment](https://github.com/woocommerce/woocommerce-ios/issues/4111#issuecomment-831909896) the Stripe Terminal SDK [supports multiple accounts and provides a mechanism for switching accounts](https://stripe.dev/stripe-terminal-ios/docs/Classes/SCPTerminal.html#/c:objc(cs)SCPTerminal(im)clearCachedCredentials).

## Changes
* Add a new action to CardPresentPaymentAction to reset the payments integration, and its implementation in CardPresentPaymentStore. This action calls disconnect and clearCacheCredentials in sequence.
* There be dragons though. `disconnect ` would crash [if called before the Terminal has been provided with a way to fetch connection tokens](https://stripe.dev/stripe-terminal-ios/docs/Classes/SCPTerminal.html#/c:objc(cs)SCPTerminal(cpy)shared). So we need to add a bit of state tracking to StripeCardReaderService.

## How to test.
Testing requires:

1. Setting up a store for Card Present Payment Testing (also P91TBi-4BH-p2 for more specific instructions)
2. The pre-release version of WCPay that implements the endpoint that captures the payment id (see p91TBi-54R#comment-4441-p2 for a downloadable zip)
3. An external hardware reader.

* Checkout the branch, build and run on device.
* Add a breakpoint to StripeCardReaderService, line 159.
<img width="350" alt="Screen Shot 2021-05-09 at 15 29 47" src="https://user-images.githubusercontent.com/2722505/117584508-6d1ee600-b0db-11eb-93ca-d23d2d09ee48.png">

* Log in, and log out immediately, without doing any operations related to payments. The breakpoint added to StripeCardReaderService should not be hit.
* Log in to a store that does not support payments. Log out or switch to a store that supports payments, the breakpoint should not be hit.
* Log in, switch to a store that supports payments. Connect to a reader. Switch stores, and the breakpoint should be hit. Switch back to the original store, attempt to pair to a reader again. It should work.
* Log in to a store that support payments. Connect to a reader. Log out. The breakpoint should be hit.
* Any variations and combinations of the above that you could think of.

Update release notes:
